managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2089,6 +2089,10 @@ void internalTrimConsumedLedgers(CompletableFuture<?> promise) {
                     log.debug("[{}] Ledger {} skipped for deletion as it is currently being written to", name,
                             ls.getLedgerId());
                     break;
+                } else if (currentLastConfirmedEntry != null && ls.getLedgerId() == currentLastConfirmedEntry.getLedgerId()) {
```

Comment:
I am not sure this is the right fix. It changes the trim behavior which causes a rolled ledger to never be deleted until a new message is produced. I think a correct fix is to fix getLastMessageId behavior to correctly handle an empty ledger case.

Reply:
@sijie thanks for checking 
Yes, I don't know what is the expected value for "lastMessageId" in this case.

It is true that this is a behaviour change in the rollover but my rationale is that if we send an "empty" messageId, the consumer (or the "Reader" in the case of the user that reported the issue) won't see a consistent sequence of "lastMessageId"
You can have:
- time 1: last messageId:   X:Y:Z:K
- time 2 (broker restart): messageId: empty
- time 3 (after one write): messageId:  X+1:Y:Z:K
so if you do not restart the broker and everything goes well you do not see that hole in the sequence

In the case of the user the lastMessageId is used to compute an approximate number of messages to be consumed by the Reader. as the topic is basically "empty" the effect is the same, that is: there are no more messages to consume. So your suggestion would make the user happy anyhow.

Regarding the trim: we are slightly changing the semantics, but as soon as you write a new message, the behavior is the same. Currently (AFAIK but I still have limited knowledge so I may be wrong) there is not strict contract about when a ledger is deleted (and also you won't know when it will be really deleted from the bookies).

All in all I would prefer this new behaviour:
- the sequence of lastMessageId does not contain "holes"
- the trim behavior changes but is it already unpredictable

You point "data won't ever be deleted if there is no more write to the topic" is valid, but I am not sure how much impact could have on production users. Is it about privacy laws/data retention rules ?

but I am open to change the patch

Reply:
in the case of the user it would be better to have this messageId working as in this patch, although we have strong rules about retention rules due to privacy laws (but you can always have a technical motivation for keeping the data for a while).

Reply:
> but I am not sure how much impact could have on production users. Is it about privacy laws/data retention rules ?

In a lot of Pulsar user cases, there are topics with relatively low traffic. If there are no writes going to the topic, it will cause accumulating one segment per topic. If there are a lot of such topics, it will cause disk filling up quickly. So I don't think we should change the retention behavior. This is a -1 from me.

As I suggested, we should fix `getLastMessageId` behavior. Because there is one empty message-id. We can still return a messageId with `entryId == -1`.

Reply:
@sijie I will adapt the patch as you are requiring.
Thanks for your feedback